### PR TITLE
feat: send btc

### PR DESCRIPTION
**Pull request by anhnh12**
#### What's this PR do?
- [x] Implement func `btcrpc.Send`

### CODE REVIEW

File: internal/blockstream/blockstream.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
+package blockstream
```
Comment (huynguyenh) [inline internal/blockstream/blockstream.go:1]: should we put it under 

internal/btcrpc/blockstream ?

File: internal/btcrpc/blockstream/blockstream.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("failed to broadcast transaction: %s", string(body))
+	}
+
+	return string(body), nil
+}
+
+func (c *blockstream) GetFeeEstimates() (map[string]float64, error) {
```
Comment (huynguyenh) [inline internal/btcrpc/blockstream/blockstream.go:57]:
```suggestion
func (c *blockstream) EstimateFee() (map[string]float64, error) {
```

Comment (huynguyenh) [inline internal/btcrpc/blockstream/blockstream.go:57]: what is the map[string]float64 stand for ?

File: internal/btcrpc/helper.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		return 0, fmt.Errorf("no fee rate available for target %d blocks", targetBlocks)
+	}
+
+	// Calculate total transaction size
+	txSize := calculateTxSize(numInputs, numOutputs)
+
+	// Calculate fee in satoshis
+	fee := int64(float64(txSize) * feeRate)
+
+	// Return minimum 1000 sats if calculated fee is too low
```
Comment (huynguyenh) [inline internal/btcrpc/helper.go:37]: why we need to do this

File: internal/btcrpc/btcrpc.go [hunk 4]
```diff
@@ -1,24 +1,182 @@
...
+	// Decode private key from WIF format
+	wif, err := btcutil.DecodeWIF(b.appConfig.Bitcoin.WalletWIF)
+	if err != nil {
+		b.logger.Fatal("failed to decode wif", map[string]string{
+			"error": err.Error(),
+		})
+	}
+	privKey := wif.PrivKey
+	pubKey := privKey.PubKey()
+	pubKeyHash := btcutil.Hash160(pubKey.SerializeCompressed())
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:44]: move this to a helper function

```getPubKey```

```func getPubKeyHash(...) ... {
   pubKey := getPubKey()
   return pubKey.Hash160(...)
}```

File: internal/btcrpc/btcrpc.go [hunk 5]
```diff
@@ -1,24 +1,182 @@
...
+	senderAddress, err := btcutil.NewAddressWitnessPubKeyHash(pubKeyHash, &chaincfg.TestNet3Params)
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:45]: move this to a helper function too `getSelfAddress`

File: internal/btcrpc/btcrpc.go [hunk 6]
```diff
@@ -1,24 +1,182 @@
...
+	if err != nil {
+		b.logger.Fatal("failed to create sender address", map[string]string{
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:47]: log Error only

File: internal/btcrpc/btcrpc.go [hunk 7]
```diff
@@ -1,24 +1,182 @@
...
+			"error": err.Error(),
+		})
+	}
+
+	// Get latest UTXO for sender address
+	latestUTXO, err := b.GetLatestUTXO(senderAddress.EncodeAddress())
+	if err != nil {
+		b.logger.Fatal("failed to get latest UTXO", map[string]string{
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:55]: log error

File: internal/btcrpc/btcrpc.go [hunk 8]
```diff
@@ -1,24 +1,182 @@
...
+			"error": err.Error(),
+		})
+	}
+
+	prevTxHash := latestUTXO.TxID
+	prevTxIndex := latestUTXO.Vout
+	prevTxAmount := latestUTXO.Value
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:62]: what's the purpose of getting this one ?

File: internal/btcrpc/btcrpc.go [hunk 9]
```diff
@@ -1,24 +1,182 @@
...
+
+	// Recipient address
+	receiverAddress, err := btcutil.DecodeAddress(receiverAddressStr, &chaincfg.TestNet3Params)
+	if err != nil {
+		b.logger.Fatal("failed to create recipient address", map[string]string{
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:67]: log err then return

File: internal/btcrpc/btcrpc.go [hunk 10]
```diff
@@ -1,24 +1,182 @@
...
+			"error":    err.Error(),
+			"receiver": receiverAddressStr,
+		})
+	}
+
+	// Transaction amounts
+	amt, ok := new(big.Int).SetString(amount.Value, 10)
+	if !ok {
+		b.logger.Fatal("failed to convert amount to big int", map[string]string{
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:76]: log err then return

File: internal/btcrpc/btcrpc.go [hunk 11]
```diff
@@ -1,24 +1,182 @@
...
+			"error":  err.Error(),
+			"amount": amount.Value,
+		})
+	}
+	amountToSend := amt.Int64()
+	txFee := int64(1000)
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:82]: why hard code fee here ?

File: internal/btcrpc/btcrpc.go [hunk 12]
```diff
@@ -1,24 +1,182 @@
...
+
+	// After verifying UTXO
+	// Calculate change amount using the actual UTXO amount
+	changeAmount := prevTxAmount - amountToSend - txFee
+	if changeAmount < 0 {
+		b.logger.Fatal("insufficient funds in UTXO")
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:88]: log err then return

File: internal/btcrpc/btcrpc.go [hunk 13]
```diff
@@ -1,24 +1,182 @@
...
+	}
+
+	// Verify the amounts add up
+	if amountToSend+txFee+changeAmount != prevTxAmount {
+		b.logger.Fatal("amount mismatch! transaction amounts don't add up to UTXO amount", map[string]string{
+			"error": err.Error(),
+		})
+	}
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:96]: move from this line upto line 73 to another smaller function `VerifyBalanceBeforeSend`

File: internal/btcrpc/btcrpc.go [hunk 14]
```diff
@@ -1,24 +1,182 @@
...
+
+	hash, err := chainhash.NewHashFromStr(prevTxHash)
+	if err != nil {
+		b.logger.Fatal("failed to create hash", map[string]string{
+			"error": err.Error(),
+		})
+	}
+
+	// Create the transaction input
+	txIn := wire.NewTxIn(wire.NewOutPoint(hash, uint32(prevTxIndex)), nil, nil)
```
Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:109]: group this into 1 smaller function `CreateTx`

Comment (huynguyenh) [inline internal/btcrpc/btcrpc.go:109]: why we create new tx with previous hash data ?

**Review by huynguyenh — CHANGES_REQUESTED** [review]

File: internal/btcrpc/blockstream/blockstream.go (hunk 2 above)
Comment (anhnh12) [reply internal/btcrpc/blockstream/blockstream.go]: - key: number of target confirmation blocks
- value: fee rates (satoshis/vB)

a btc tx usually targets for 6 confirmation blocks

File: internal/btcrpc/btcrpc.go (hunk 8 above)
Comment (anhnh12) [reply internal/btcrpc/btcrpc.go]: UTXO: Unspent tx output
A bitcoin transfer needs previous txs data as inputs. They can be used as proof to verify source of the btc amount you are trying to transfer.
e.g. transfer 1000 sats, need to refer to one or more previous txs where the sum of UTXOs amount >= 1000 sats

How it works
transfer amount = 1000 sats, use tx A as UTXO (1200 sats)
-> this UTXO is marked as spent, can no longer be used in other txs
-> 1200 spent -> changeAmount = 200 -> so we got a new UTXO as 200 sats

**Pull request merged**